[docs] Filters (#2686) (#70)
OpenCTI-Platform/docs: docs/usage/filters.md

```
FilterGroup: {
    mode: 'and' | 'or',
    filters: Filter[],
    filterGroups: FilterGroup[],
}

Filter {
    key: string,
    values: string[],
    operator: string,
    mode: 'and' | 'or',
}
```

In a filter group, the 'mode' can be equal to 'and' or 'or' and represent the operator between the different filters and groups in the 'filters' and 'filterGroups' arrays.
The 'filters' and 'filterGroups' arrays are composed of objects of type Filter/FilterGroup.
The base entity of a filter group object is a filter with 4 attributes:
- a key, representing on what we want to do the filtering (exemple: 'objectLabel' to filter on labels, 'createdBy' to filter on the entity creation date, 'report_types', 'entity_type'... ).
- an array of values, representing the values we are interested in.
- an operator representing the operation we want to apply between the key and the values.
- a mode ('and' or 'or') to apply between the values if there are several ones.

### Operators
The available operators are:
- ```'eq'```  for 'equal',
  

    Exemple: ``entity_type = 'Report'``

- ```'not_eq'``` for 'unequal',
- ```'gt'/'gte'``` for 'greater than'/'greater than or equal',

>     Exemple: ``confidence > 50``

Reviewer: I don't understand why there is a `gt` operator and a `>` in the example. is it to enlight that `>` is represented by `gt` in the filter code ?

Author: right that's a mistake. I'm going to fix this.

- ```'lt'/'lte'``` for 'lower than'/'lower than or equal',
- ```'nil'``` for 'empty',

    Exemple: ``marking is empty``

- ```'not_nil'``` for 'not empty'.

> 

Reviewer: ```suggestion
Note: some operator may not be allowed for some key, for additional information please navigate to the "Special keys" section.
```

> ### Examples

Reviewer: ```suggestion
### Examples of filter using OpenCTI version 5.12+
```
this is my understanding, feel free to correct me

- entity_type = 'Report'
```
filters = {
    mode: 'and',
    filters: [{
        key: 'entity_type',
        values: ['Report'],
        operator: 'eq',
        mode: 'or', // useless here because values contains only 1 value
    }],
    filterGroups: [],
};
```

- (entity_type = 'Report') AND (label = 'label1' OR 'label2')
```
filters = {
    mode: 'and',
    filters: [
        {
            key: 'entity_type',
            values: ['Report'],
            operator: 'eq',
            mode: 'or',
        },
        {
            key: 'objectLabel',
            values: ['label1', 'label2'],
            operator: 'eq',
            mode: 'or',
        }
    ],
    filterGroups: [],
};
```

> - (entity_type = 'Report') AND (confidence > 50 OR marking is empty)

Reviewer: ❤️ this example is very explicit for me

#### Supported filter keys in streams and triggers
In streams and triggers, only some keys are supported for the moment:
- ``objectAssignee``
- ``confidence``
- ``createdBy``
- ``creator``
- ``x_opencti_detection``
- ``indicator_types``
- ``objectLabel``
- ``x_opencti_main_observable_type``
- ``objectMarking``
- ``objects``
- ``pattern_type``
- ``priority``
- ``revoked``
- ``severity``
- ``x_opencti_score``
- ``entity_type``
- ``x_opencti_workflow_id``
- ``connectedToId`` (for instance triggers)
- ``fromId`` (the instance in the 'from' of a relationship)
- ``fromTypes`` (the entity type in the 'from' of a relationship)
- ``toId``
- ``toTypes``

#### Some important keys
Here are some important keys as an example. For a complete list of the available keys, see the attributes and relations schema definitions.
X refers here to the filtered entities.
- ``objectLabel``: label of X,
- ``objectMarking``: marking of X,
- ``createdBy``: author of X,
- ``creator_id``: technical creator of X,
- ``created_at``: date of creation of X in the platform,
- ``confidence``: confidence of X,
- ``entity_type``: X entity type ('Report', 'Stix-Cyber-Observable', ...),

#### Special keys

> Some keys are a bit special: there are not existing in the schema definition, but are allowed. They describe a special behavior.

Reviewer: ```suggestion
Some keys does not exist in the schema definition, but are allowed in addition. They describe a special behavior.
```

# Filter knowledge

In OpenCTI, you can filter data to target or view entities that have particular attributes.

## Filters usages
Filters are used in different locations in the platform.
- in entities lists: to display only the entities matching the filters. If an export or a background task is generated, only the filtered data will be taken into account.
- in investigations and knowledge graphs: to display only the entities matching the filters.

> - in workspaces: to create graphs with only the entities matching the filters.

Reviewer 2: ```suggestion
- in Dashboards: to create Widget graphs and lists with only the entities matching the filters.
```

Reviewer 2: Workspace has no meaning for users.

# Filter knowledge

In OpenCTI, you can filter data to target or view entities that have particular attributes.

## Filters usages
Filters are used in different locations in the platform.
- in entities lists: filters restrict the displayed entities to the ones respecting the chosen filtering. If an export or a background task is generated, only the filtered data will be taken into account.
- in investigations and knowledge graphs: filters restrict the displayed entities to the ones respecting the chosen filtering.
- in workspaces: filters enable to create graphs with only the wanted entities.
- in feeds, taxii collections, triggers, streams, playbooks: filters restrict the data to manage or the events to listen to.

There are two types of filters:
- dynamic filters: They are not stored in the database, they enable to filter view in the UI. Exemples: filters in entities list, investigations, knowledge graphs;

> - stored filters: They are attributes of an entity, they are stored as they other attributes. Exemples: filters of workspaces, feeds, taxii collections, triggers, streams, playbooks.

Author: right! this formulation is better :)

> - stored filters: They are attributes of an entity, they are stored as they other attributes. Exemples: filters of workspaces, feeds, taxii collections, triggers, streams, playbooks.

Author: they are internal object ;)

> ## Filters format (since 5.12)

Reviewer 3: We could split in 2 sections : UI and API usages. 
This section here is the API.

Author: right!

The filters format since OpenCTI 5.12 is of type FilterGroup (see below). This is the format to use when doing query via the API.
This enables to do complex filters imbrication with different operators ('and', 'or').

```
FilterGroup: {
    mode: string, // 'and' or 'or'
    filters: Filter[],
    filterGroups: FilterGroup[],
}

Filter {
    key: string,
    values: string[],
    operator: string,
    mode: string, // 'and' or 'or'
}
```

In a filter group, the 'mode' can be equal to 'and' or 'or' and represent the operator between the different filters and groups in the 'filters' and 'filterGroups' arrays.
The 'filters' and 'filterGroups' arrays are composed of objects of type Filter/FilterGroup.
The base entity of a filter group object is a filter with 4 attributes:

> - a key, representing on what we want to do the filtering (exemple: 'objectLabel' to filter on labels, 'createdBy' to filter on the entity creation date, 'report_types', 'entity_type'... ).

Reviewer 3: if we want a complete API documentation, we need an exhaustive list. Some if these keys are not trivial 😉

Author: I'm gonna list some important keys + keys with a special behavior

The former format was not adapted for such complex filters, and a refacto was necessary.
Indeed:
- the filters frontend and backend formats were different, requiring conversions,
- the filter keys were enums, i.e. the keys should belong to the static list of the available keys for the given entity type,
- in the frontend format, the operator was contained in the key, limiting the operators combinations and creating useless complexity of parsing,
- the frontend format didn't contained a mode attribute, avoiding modes (and/or) customization,
- the flat structure didn't enable filters imbrication at several levels.

```
// old filters format
FrontendFilters: Map<string, { id: string | null, value: string }>

BackendFilters {
    key: string, // an enum that should be in the list of the available filter keys for given the entity type
    values: string[],
    operator: string,
    filterMode: string, // 'and' or 'or'
}
```

Because changing filters format impacts almost everything in the platform, we decided to do a complete refacto once and for all, a refacto that would enable almost all complex combinations and compatible with our long term vision of filters in OpenCTI.

> ### Content: what has been done

Author: yes! the migration part can be put in a blog post + in the release not , or in the release warning , or elsewhere

> ### Content: what has been done

Author: yes, this doc is API centered (for UI, waiting for the UI to be done)... we'll see with Ju what to do on Friday, for the moment, I let the doc here to facilitate reviews
